File: backend/models/Content.js
// models/Content.js
const mongoose = require('mongoose');

const contentSchema = new mongoose.Schema({
    // 基本信息
    title: {
        'zh-CN': { type: String, required: true },
        'en-US': { type: String },
        'ja-JP': { type: String },
        'ko-KR': { type: String }
    },
    description: {
        'zh-CN': { type: String },
        'en-US': { type: String },
        'ja-JP': { type: String },
        'ko-KR': { type: String }
    },
    
    // 内容类型
    contentType: {
        type: String,
        enum: ['180_video', '180_photo', '360_video', '360_photo', 'spatial_video', 'spatial_photo'],
        required: true
    },
    
    // 文件信息
    files: {
        main: {
            url: { type: String, required: true },
            size: Number,
            duration: Number, // 视频时长（秒）
            resolution: {
                width: Number,
                height: Number
            }
        },
        thumbnail: {
            url: String,
            size: Number
        },
        preview: {
            url: String,
            size: Number
        }
    },
    
    // 额外媒体
    media: {
        backgroundMusic: {
            url: String,
            title: String,
            artist: String
        },
        narration: {
            'zh-CN': { url: String },
            'en-US': { url: String },
            'ja-JP': { url: String },
            'ko-KR': { url: String }
        }
    },
    
    // GPS信息
    location: {
        latitude: Number,
        longitude: Number,
        altitude: Number,
        address: String,
        country: String,
        city: String
    },
    
    // 创作者信息
    creatorId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    
    // 协作信息
    collaboration: {
        isCollaborative: { type: Boolean, default: false },
        collaborationId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'Collaboration'
        }
    },
    
    // 标签和分类
    tags: [String],
    category: {
        type: String,
        enum: ['travel', 'education', 'entertainment', 'sports', 'news', 'documentary', 'art', 'other'],
        default: 'other'
    },
    
    // 付费设置
    pricing: {
        isFree: { type: Boolean, default: true },
        price: { type: Number, default: 0 },
        currency: { type: String, default: 'CNY' }
    },
    
    // 统计信息
    stats: {
        views: { type: Number, default: 0 },
        likes: { type: Number, default: 0 },
        favorites: { type: Number, default: 0 },
        comments: { type: Number, default: 0 },
        shares: { type: Number, default: 0 },
        downloads: { type: Number, default: 0 },
        danmakus: { type: Number, default: 0 }
    },
    // 内容状态
    status: {
        type: String,
        enum: ['draft', 'pending_review', 'approved', 'rejected', 'published', 'archived'],
        default: 'draft'
    },
    
    // 审核信息
    review: {
        reviewedBy: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User'
        },
        reviewedAt: Date,
        reviewNote: String
    },
    
    // 版本控制
    version: {
        current: { type: Number, default: 1 },
        history: [{
            version: Number,
            changedBy: {
                type: mongoose.Schema.Types.ObjectId,
                ref: 'User'
            },
            changedAt: Date,
            changes: mongoose.Schema.Types.Mixed
        }]
    },
    
    // 时间戳
    publishedAt: Date,
    createdAt: { type: Date, default: Date.now },
    updatedAt: { type: Date, default: Date.now }
}, {
    timestamps: true
});

// 索引
contentSchema.index({ creatorId: 1, status: 1 });
contentSchema.index({ 'title.zh-CN': 'text', 'description.zh-CN': 'text' });
contentSchema.index({ tags: 1 });
contentSchema.index({ category: 1 });
contentSchema.index({ 'location.latitude': 1, 'location.longitude': 1 });

// 虚拟字段：获取本地化标题
contentSchema.virtual('localizedTitle').get(function() {
    return (language) => {
        return this.title[language] || this.title['zh-CN'];
    };
});

// 实例方法：更新统计
contentSchema.methods.incrementStat = async function(statType) {
    if (this.stats[statType] !== undefined) {
        this.stats[statType] += 1;
        await this.save();
    }
};

// 实例方法：添加版本记录
contentSchema.methods.addVersionHistory = async function(userId, changes) {
    const newVersion = this.version.current + 1;
    
    this.version.history.push({
        version: this.version.current,
        changedBy: userId,
        changedAt: new Date(),
        changes: changes
    });
    
    this.version.current = newVersion;
    await this.save();
};

module.exports = mongoose.model('Content', contentSchema);

File: VisionProApp/VisionProApp/Models/Content.swift
//
//  Content.swift
//  VisionProApp
//
//  Created by 五行 on 2025/5/11.
//
// Models/Content.swift
import Foundation

// 创作者信息
struct Creator: Codable {
    let id: String
    let username: String
    let email: String?
    let profile: Profile?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case username
        case email
        case profile
    }
}

struct Profile: Codable {
    let displayName: [String: String]?
    let bio: [String: String]?
}

struct Content: Codable, Identifiable, Hashable {
    let id: String
    let title: [String: String]
    let description: [String: String]?
    let contentType: ContentType
    let files: ContentFiles
    let media: ContentMedia?
    let location: Location?
    let creatorId: Creator
    let tags: [String]?
    let category: ContentCategory
    let pricing: Pricing
    let stats: ContentStats
    let status: ContentStatus
    let thumbnailURL: String?
    let createdAt: Date?
    let updatedAt: Date?
    
    enum ContentType: String, Codable, CaseIterable {
        case video180 = "180_video"
        case photo180 = "180_photo"
        case video360 = "360_video"
        case photo360 = "360_photo"
        case spatialVideo = "spatial_video"
        case spatialPhoto = "spatial_photo"
    }
    
    enum ContentCategory: String, Codable, CaseIterable {
        case travel = "travel"
        case education = "education"
        case entertainment = "entertainment"
        case sports = "sports"
        case news = "news"
        case documentary = "documentary"
        case art = "art"
        case other = "other"
    }
    
    enum ContentStatus: String, Codable {
        case draft = "draft"
        case pendingReview = "pending_review"
        case approved = "approved"
        case rejected = "rejected"
        case published = "published"
        case archived = "archived"
    }
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case title
        case description
        case contentType
        case files
        case media
        case location
        case creatorId
        case tags
        case category
        case pricing
        case stats
        case status
        case thumbnailURL
        case createdAt
        case updatedAt
    }
    
    // Hashable
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    static func == (lhs: Content, rhs: Content) -> Bool {
        lhs.id == rhs.id
    }
}

struct ContentFiles: Codable {
    let main: MainFile
    let thumbnail: FileInfo?
    let preview: FileInfo?
}

struct MainFile: Codable {
    let url: String
    let size: Int?
    let duration: Int?
    let resolution: Resolution?
}

struct FileInfo: Codable {
    let url: String
    let size: Int?
}

struct Resolution: Codable {
    let width: Int
    let height: Int
}

struct ContentMedia: Codable {
    let backgroundMusic: BackgroundMusic?
    let narration: [String: Narration]?
}

struct BackgroundMusic: Codable {
    let url: String
    let title: String?
    let artist: String?
}

struct Narration: Codable {
    let url: String
}

struct Location: Codable {
    let latitude: Double
    let longitude: Double
    let altitude: Double?
    let address: String?
    let country: String?
    let city: String?
}

struct Pricing: Codable {
    let isFree: Bool
    let price: Double?
    let currency: String?
}

struct ContentStats: Codable {
    let views: Int
    let likes: Int
    let favorites: Int
    let comments: Int
    let shares: Int
    let downloads: Int
    let danmakus: Int
}

// 扩展用于便捷访问
extension Content {
    var localizedTitle: String {
        let lang = Locale.current.language.languageCode?.identifier ?? "zh-CN"
        return title[lang] ?? title["zh-CN"] ?? ""
    }
    
    var localizedDescription: String {
        let lang = Locale.current.language.languageCode?.identifier ?? "zh-CN"
        return description?[lang] ?? description?["zh-CN"] ?? ""
    }
    
    var isPaid: Bool {
        return !pricing.isFree
    }
    
    var formattedPrice: String {
        if pricing.isFree {
            return "免费"
        }
        let price = pricing.price ?? 0
        let currency = pricing.currency ?? "CNY"
        return "\(currency) \(String(format: "%.2f", price))"
    }
    
    var isVideo: Bool {
        return contentType == .video180 || contentType == .video360 || contentType == .spatialVideo
    }
    
    var isPhoto: Bool {
        return contentType == .photo180 || contentType == .photo360 || contentType == .spatialPhoto
    }
}
